mlflow/sklearn/__init__.py

```diff
@@ -1612,6 +1651,31 @@ def patched_predict(original, self, *args, **kwargs):
             _AUTOLOGGING_METRICS_MANAGER.register_prediction_result(
                 run_id, eval_dataset_name, predict_result
             )
+            if log_datasets:
+                try:
+                    context_tags = context_registry.resolve_tags()
+                    source = CodeDatasetSource(
+                        mlflow_source_type=context_tags[MLFLOW_SOURCE_TYPE],
+                        mlflow_source_name=context_tags[MLFLOW_SOURCE_NAME],
+                    )
+
+                    dataset = _create_dataset(eval_dataset, source, dataset_name=eval_dataset_name)
+
+                    # log the dataset
+                    if dataset:
+                        tags = [InputTag(key=MLFLOW_DATASET_CONTEXT, value="eval")]
+                        dataset_input = DatasetInput(dataset=dataset._to_mlflow_entity(), tags=tags)
+
+                        # log the dataset
+                        client = mlflow.MlflowClient()
+                        client.log_inputs(
+                            run_id=mlflow.active_run().info.run_id, datasets=[dataset_input]
```

Comment:
Looks like `mlflow.active_run()` might not always be the right thing to use for evaluation. I tried the following code, and it failed to log the evaluation dataset with the following error:

```
import mlflow.sklearn

mlflow.sklearn.autolog()

from sklearn.linear_model import LogisticRegression
from scipy.stats import uniform

logistic = LogisticRegression(solver="saga", tol=1e-2, max_iter=200, random_state=0)
distributions = dict(C=uniform(loc=0, scale=4), penalty=["l2", "l1"])

from sklearn import datasets

iris = datasets.load_iris()

logistic.fit(iris.data, iris.target)

import sklearn.base

import numpy as np

print(logistic.predict(np.array([[5.1, 3.5, 1.4, 0.2], [4.9, 3, 1.4, 0.2]])))
```

```
python testdouble.py
2023/04/25 15:19:20 INFO mlflow.utils.autologging_utils: Created MLflow autologging run with ID 'a00820a4360d447d8c533c56cdfe397e', which will track hyperparameters, performance metrics, model artifacts, and lineage information for the current sklearn workflow
2023/04/25 15:19:27 WARNING mlflow.sklearn: Failed to log evaluation dataset information to MLflow Tracking. Reason: 'NoneType' object has no attribute 'info'
[0 0]
```

Can we fix this for this flavor and others? Can we also add test cases?

Reply:
^ I think this happens because the run ends when `logistic.fit()` terminates. We should see how `logistic.predict()` currently gets the run ID - I think it references some global state that we capture during `logistic.fit()` :)

Reply:
Ah good catch. Added a test case `sklearn` and `pyspark`, which are the two flavors that do autologging in the `model.predict()`/`model.evaluate()` apis